### PR TITLE
2021/12/11 Multiprocessing

### PR DESCRIPTION
**Pull request by jotelha**
This outsources the dtool API's copy function to another process and receives status updates. The mock utils.processbar.ProcessBar replaces the click.ProcessBar expected by dtoolcore. Right now, only prints log messages. Should be directly attachable to a Gtk.ProgressBar. The utils.muliprocessing utility module needs a bit of cleaning up.

For a simple demonstration, 

```python
from dtool_lookup_gui.utils.multiprocessing import test_run
test_run()
```

will produce output like

```
Child process step 0
Child process step 1
Child process step 2
Child process step 3
Test callback received report for step 0
Test callback received report for step 1
Child process step 4
Child process step 5
Test callback received report for step 2
Child process step 6
Test callback received report for step 3
Child process step 7
Test callback received report for step 4
Child process step 8
Test callback received report for step 5
Child process step 9
Test callback received report for step 6
Test callback received report for step 7
Test callback received report for step 8
Test callback received report for step 9
Child process returned True.
```

### CODE REVIEW

File: dtool_lookup_gui/utils/multiprocessing.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        """Spawn child process to assure my environment stays untouched."""
+        return_value_queue = multiprocessing.Queue()
+        status_progress_queue = multiprocessing.Queue()
+        stop_event = multiprocessing.Event()
+        process = Process(target=self.target_wrapper, args=[return_value_queue, status_progress_queue, stop_event, *args])
+        process.start()
+
+        # wait for child to queue objects and
+        # check whether child raises exception
+        while return_value_queue.empty():
```
Comment (pastewka) [inline dtool_lookup_gui/utils/multiprocessing.py:100]: I don't understand this. Doesn't this while loop block?

**Comment by pastewka** [comment]
I think I don't understand what this module does and would like to discuss. In particular, I don't see how this ties into `asyncio` and leads to a non-blocking call. You need to poll the queue to get those status updates.

File: dtool_lookup_gui/models/datasets.py [hunk 2]
```diff
@@ -182,12 +187,22 @@ def _copy_dataset(uri, target_base_uri, resume, auto_resume):
                 raise FileExistsError(
                     "Path already exists: {}".format(parsed_dataset_uri.path))
 
-    dest_uri = copy_func(
-        src_uri=uri,
-        dest_base_uri=target_base_uri,
-        config_path=None
-        #progressbar=progressbar
-    )
+    def copy_func_wrapper(src_uri, dest_base_uri, status_report_callback, stop_event_callback):
```
[the remark this answers is not in the document]
Comment (jotelha) [reply dtool_lookup_gui/models/datasets.py:190]: The `StatusReportingChildProcessBuilder` expects the keyword arguments `status_report_callback` and `stop_event_callback`, hence. Will change that and remove the need for this wrapper.

File: dtool_lookup_gui/utils/multiprocessing.py (hunk 1 above)
Comment (jotelha) [reply dtool_lookup_gui/utils/multiprocessing.py:100]: It shouldn't, the only blocking lines should be those at the bottom, https://github.com/IMTEK-Simulation/dtool-lookup-gui/blob/bd0e9b33040316749f7db1d3f3e36f0b610f482e/dtool_lookup_gui/utils/multiprocessing.py#L125-L128.

Comment (pastewka) [reply dtool_lookup_gui/utils/multiprocessing.py:100]: Don't you need to poll the queue regularly to get the progress information? How does that work?

**Comment by jotelha** [comment]
My apologies, last evening I was quite happy that the thing seemed to actually work without much effort and overlooked that it just did not block because the call had still been wrapped in this `ProcessPoolExecutor` context, https://github.com/IMTEK-Simulation/dtool-lookup-gui/blob/bd0e9b33040316749f7db1d3f3e36f0b610f482e/dtool_lookup_gui/models/datasets.py#L303-L304 :| The while loop sure was blocking, now it's not anymore.

File: dtool_lookup_gui/utils/multiprocessing.py (hunk 1 above)
Comment (jotelha) [reply dtool_lookup_gui/utils/multiprocessing.py:96]: There are two queues, one for the progress and one only for the return value of the function (if it has any). The two polling calls within the loop https://github.com/IMTEK-Simulation/dtool-lookup-gui/blob/f55e6bca037b9a179070308f2de20999a09a3e29/dtool_lookup_gui/utils/multiprocessing.py#L101 and https://github.com/IMTEK-Simulation/dtool-lookup-gui/blob/f55e6bca037b9a179070308f2de20999a09a3e29/dtool_lookup_gui/utils/multiprocessing.py#L109 don't wait. The loop itself of course was blocking, sorry, see below ...

Comment (pastewka) [reply dtool_lookup_gui/utils/multiprocessing.py:96]: Okay, now it makes sense

**Review by pastewka — APPROVED** [review]

**Pull request merged**